File: app/src/main/java/io/github/lee0701/heonot/InputMethodSettingsActivity.java
package io.github.lee0701.heonot;

import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.util.TypedValue;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import org.json.JSONException;

import io.github.lee0701.heonot.inputmethod.InputMethod;
import io.github.lee0701.heonot.inputmethod.modules.InputMethodModule;

public class InputMethodSettingsActivity extends SettingsActivity {

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_input_method_settings);

		Intent intent = getIntent();
		InputMethod method = inputMethods.get(intent.getIntExtra(EXTRA_METHOD_ID, -1));

		LinearLayout linearLayout = (LinearLayout) findViewById(R.id.modules);
		linearLayout.setOrientation(LinearLayout.VERTICAL);

		for(InputMethodModule module : method.getModules()) {
			TextView textView = new TextView(this);
			textView.setTextSize(TypedValue.COMPLEX_UNIT_PT, 12);
			textView.setText(module.getName());
			linearLayout.addView(textView);
			linearLayout.addView(module.createSettingsView(this));
		}

		RelativeLayout relativeLayout = new RelativeLayout(this);

		Button exportButton = new Button(this);
		exportButton.setText(R.string.button_export);
		exportButton.setOnClickListener((v) -> {
			String json;
			try {
				json = method.toJSON(2);
			} catch(JSONException e) {
				return;
			}
			new AlertDialog.Builder(this)
					.setMessage(R.string.msg_copy_method_json)
					.setPositiveButton(R.string.button_ok, (dialog, which) -> {
						ClipboardManager clipboard = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
						if(Build.VERSION.SDK_INT > 11) {
							ClipData clipData = ClipData.newPlainText("JSON Data", json);
							clipboard.setPrimaryClip(clipData);
						} else {
							clipboard.setText(json);
						}
					})
					.setNegativeButton(R.string.button_cancel, (dialog, which) -> {})
					.create()
			.show();
		});
		relativeLayout.addView(exportButton);

		linearLayout.addView(relativeLayout);

		Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
		setSupportActionBar(toolbar);

	}

}

File: app/src/main/java/io/github/lee0701/heonot/InputMethodSettingsActivity.kt
package io.github.lee0701.heonot

import android.app.AlertDialog
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.widget.Button
import android.widget.LinearLayout
import io.github.lee0701.heonot.inputmethod.InputMethod
import kotlinx.android.synthetic.main.activity_input_method_settings.*
import kotlinx.android.synthetic.main.settings_toolbar.*
import org.json.JSONException


class InputMethodSettingsActivity : SettingsActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_input_method_settings)

        val id = intent.getIntExtra(SettingsActivity.EXTRA_METHOD_ID, -1)
        var method = globalModules
        if(id >= 0) method = inputMethods[id]

        for (module in method.modules) {
            modules.addTab(modules.newTab().setText(module.name))
        }

        val listener = object: TabLayout.OnTabSelectedListener {
            override fun onTabReselected(tab: TabLayout.Tab?) {}
            override fun onTabUnselected(tab: TabLayout.Tab?) {}
            override fun onTabSelected(tab: TabLayout.Tab?) {
                module_settings.removeAllViews()
                module_settings.addView(method.modules.get(modules.selectedTabPosition).createSettingsView(this@InputMethodSettingsActivity))

            }
        }
        modules.addOnTabSelectedListener(listener)

        val importButton = Button(this@InputMethodSettingsActivity)
        importButton.setText(R.string.button_import)
        importButton.setOnClickListener {
            AlertDialog.Builder(this@InputMethodSettingsActivity)
                    .setMessage(R.string.msg_paste_method_json)
                    .setPositiveButton(R.string.button_ok) { _, _ ->
                        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                        val json = run {
                            if (Build.VERSION.SDK_INT >= 11) {
                                val clipData = clipboard.primaryClip
                                clipData.getItemAt(0)?.coerceToText(this@InputMethodSettingsActivity).toString()
                            } else {
                                clipboard.text.toString()
                            }
                        }
                        try {
                            if(json.isNotEmpty()) {
                                val method = InputMethod.loadJSON(json)
                                inputMethods[id] = method
                                saveMethods()
                                finish()
                            } else throw Exception()
                        } catch(ex: Exception) {
                            AlertDialog.Builder(this@InputMethodSettingsActivity)
                                    .setMessage(R.string.msg_error_importing)
                                    .create()
                                    .show()
                        }
                    }
                    .setNegativeButton(R.string.button_cancel, { _, _ -> })
                    .create()
                    .show()
        }
        buttons.addView(importButton)

        val exportButton = Button(this@InputMethodSettingsActivity)
        exportButton.setText(R.string.button_export)
        exportButton.setOnClickListener { _ ->
            val json: String
            try {
                json = method.toJSON(2)
            } catch (e: JSONException) {
                return@setOnClickListener
            }

            AlertDialog.Builder(this@InputMethodSettingsActivity)
                    .setMessage(R.string.msg_copy_method_json)
                    .setPositiveButton(R.string.button_ok) { _, _ ->
                        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                        if (Build.VERSION.SDK_INT >= 11) {
                            val clipData = ClipData.newPlainText("JSON Data", json)
                            clipboard.primaryClip = clipData
                        } else {
                            clipboard.text = json
                        }
                    }
                    .setNegativeButton(R.string.button_cancel, { _, _ -> })
                    .create()
                    .show()
        }
        buttons.addView(exportButton)

        val duplicateButton = Button(this@InputMethodSettingsActivity)
        duplicateButton.setText(R.string.button_duplicate)
        duplicateButton.setOnClickListener({ _ ->
            inputMethods.add(InputMethod(method))
            saveMethods()
            finish()
        })
        buttons.addView(duplicateButton)

        val deleteButton = Button(this@InputMethodSettingsActivity)
        deleteButton.setText(R.string.button_delete)
        deleteButton.setOnClickListener { _ ->
            inputMethods.remove(method)
            saveMethods()
            finish()
        }
        buttons.addView(deleteButton)

        listener.onTabSelected(modules.getTabAt(0))

        setSupportActionBar(toolbar)

    }
}
